Trampoline Forwarding: Enforce trampoline constraints

a-mpch opened the pull request:
This PR is part of 1/N PRs in order to split up #3976

 This commit checks amount and CLTV of the trampoline to the outer hop. If exceeds limitations we fail the forward using a `InboundHTLCErr` with specified local reason.

tankyleo commented on lightning/src/ln/onion_utils.rs:
```diff
@@ -1718,6 +1725,9 @@ impl LocalHTLCFailureReason {
 			Self::InvalidOnionPayload | Self::InvalidTrampolinePayload => PERM | 22,
 			Self::MPPTimeout => 23,
 			Self::InvalidOnionBlinding => BADONION | PERM | 24,
+			Self::TemporaryTrampolineFailure => PERM | 25,
+			Self::TrampolineFeeOrExpiryInsufficient => PERM | 26,
```
Both of these should be `NODE` I think, not `PERM`

tankyleo commented on lightning/src/ln/onion_utils.rs:
```diff
@@ -2018,6 +2031,11 @@ impl HTLCFailReason {
 			},
+			LocalHTLCFailureReason::TemporaryTrampolineFailure => debug_assert!(data.is_empty()),
+			LocalHTLCFailureReason::TrampolineFeeOrExpiryInsufficient => {
+				debug_assert!(data.is_empty())
```
should be `debug_assert_eq!(data.len(), 10)`

tankyleo commented on lightning/src/ln/onion_payment.rs:
```diff
@@ -135,7 +154,25 @@ pub(super) fn create_fwd_pending_htlc_info(
+						outer_hop_data.outgoing_cltv_value.write(&mut err_data).unwrap();
+					}
+					LocalHTLCFailureReason::FinalIncorrectHTLCAmount => {
+						outer_hop_data.amt_to_forward.write(&mut err_data).unwrap();
```
Should this write `mtd.total_msat` if we have multipath trampoline data ?

tankyleo commented on lightning/src/ln/onion_payment.rs:
```diff
@@ -281,14 +327,35 @@ pub(super) fn create_recv_pending_htlc_info(
+						outer_hop_data.outgoing_cltv_value.write(&mut err_data).unwrap();
+					}
+					LocalHTLCFailureReason::FinalIncorrectHTLCAmount => {
+						outer_hop_data.amt_to_forward.write(&mut err_data).unwrap();
```
Similar question about writing `mtd.total_msat`

tankyleo commented on lightning/src/ln/onion_payment.rs:
```diff
@@ -135,7 +154,25 @@ pub(super) fn create_fwd_pending_htlc_info(
+				InboundHTLCErr {
+					reason,
+					err_data,
+					msg: "Underflow calculating outbound amount or CLTV value for Trampoline forward",
```
I don't find this very clear here and elsewhere :) I would have written something in the style of the comment just above here
```
Trampoline onion's amt or CTLV values exceeded the outer onion's
```

tankyleo commented on lightning/src/ln/onion_payment.rs:
```diff
@@ -135,7 +154,25 @@ pub(super) fn create_fwd_pending_htlc_info(
-		onion_utils::Hop::TrampolineForward { next_trampoline_hop_data, next_trampoline_hop_hmac, new_trampoline_packet_bytes, trampoline_shared_secret, .. } => {
+		onion_utils::Hop::TrampolineForward { ref outer_hop_data, next_trampoline_hop_data, next_trampoline_hop_hmac, new_trampoline_packet_bytes, trampoline_shared_secret, .. } => {
+			check_trampoline_onion_constraints(outer_hop_data, next_trampoline_hop_data.outgoing_cltv_value, next_trampoline_hop_data.amt_to_forward).map_err(|reason| {
+				let mut err_data = Vec::new();
+				match reason {
+					LocalHTLCFailureReason::FinalIncorrectCLTVExpiry => {
+						outer_hop_data.outgoing_cltv_value.write(&mut err_data).unwrap();
+					}
+					LocalHTLCFailureReason::FinalIncorrectHTLCAmount => {
+						outer_hop_data.amt_to_forward.write(&mut err_data).unwrap();
+					}
+					_ => unreachable!()
+				}
```
We can remove the trait import and do this here and below:
```diff
diff --git a/lightning/src/ln/onion_payment.rs b/lightning/src/ln/onion_payment.rs
index 91bd2cce2..e7db246a2 100644
--- a/lightning/src/ln/onion_payment.rs
+++ b/lightning/src/ln/onion_payment.rs
@@ -22,7 +22,6 @@ use crate::sign::{NodeSigner, Recipient};
 use crate::types::features::BlindedHopFeatures;
 use crate::types::payment::PaymentHash;
 use crate::util::logger::Logger;
-use crate::util::ser::Writeable;

 #[allow(unused_imports)]
 use crate::prelude::*;
@@ -158,17 +157,17 @@ pub(super) fn create_fwd_pending_htlc_info(
                        }),
                onion_utils::Hop::TrampolineForward { ref outer_hop_data, next_trampoline_hop_data, next_trampoline_hop_hmac, new_trampoline_packet_bytes, trampoline_shared_secret, .. } => {
                        check_trampoline_onion_constraints(outer_hop_data, next_trampoline_hop_data.outgoing_cltv_value, next_trampoline_hop_data.amt_to_forward).map_err(|reason| {
-                               let mut err_data = Vec::new();
+                               let err_data;
                                match reason {
                                        LocalHTLCFailureReason::FinalIncorrectCLTVExpiry => {
-                                               outer_hop_data.outgoing_cltv_value.write(&mut err_data).unwrap();
+                                               err_data = outer_hop_data.outgoing_cltv_value.to_be_bytes().to_vec();
                                        }
                                        LocalHTLCFailureReason::FinalIncorrectHTLCAmount => {
                                                let outgoing_amount = outer_hop_data
                                                        .multipath_trampoline_data
                                                        .as_ref()
                                                        .map_or(outer_hop_data.amt_to_forward, |mtd| mtd.total_msat);
-                                               outgoing_amount.write(&mut err_data).unwrap();
+                                               err_data = outgoing_amount.to_be_bytes().to_vec();
                                        }
                                        _ => unreachable!()
                                }
@@ -340,13 +339,14 @@ pub(super) fn create_recv_pending_htlc_info(
                        }, ..
                } => {
                        check_trampoline_onion_constraints(outer_hop_data, cltv_expiry_height, sender_intended_htlc_amt_msat).map_err(|reason| {
-                               let mut err_data = Vec::new();
+                               let err_data;
                                match reason {
                                        LocalHTLCFailureReason::FinalIncorrectCLTVExpiry => {
-                                               outer_hop_data.outgoing_cltv_value.write(&mut err_data).unwrap();
+                                               err_data = outer_hop_data.outgoing_cltv_value.to_be_bytes().to_vec();
                                        }
                                        LocalHTLCFailureReason::FinalIncorrectHTLCAmount => {
-                                               outer_hop_data.amt_to_forward.write(&mut err_data).unwrap();
+                                               err_data = outer_hop_data.amt_to_forward.to_be_bytes().to_vec();
+
                                        }
                                        _ => unreachable!()
                                }
```

tankyleo reviewed (commented):
This is a more mechanical review as I am not familiar with this part of the codebase :)

tankyleo commented:
Also when I remove everything but the test code, the test does *not* fail - is that intentional ? I have not reviewed the test

a-mpch commented:
thanks @tankyleo !

> Also when I remove everything but the test code, the test does _not_ fail - is that intentional ? I have not reviewed the test

Double checked the tests and yes, this happens because when receiving a trampoline it behaves exactly as receiving a non-trampoline payment. Having the same failure scenario.
I figured out that for this PR it's ok to have one test that checks for the CLTV constraint and later when having the capacity of forwarding we can test the constraint when forwarding related to the amt.

I added fixups for each of your comments.

tankyleo commented on lightning/src/ln/blinded_payment_tests.rs:
```diff
@@ -2288,8 +2287,8 @@ fn test_trampoline_unblinded_receive() {
-						fee_msat: amt_msat,
-						cltv_expiry_delta: 24,
+						fee_msat: 0,
+						cltv_expiry_delta: 72,
```
Similar here about why this is needed thank you (take a look at this comment on the original commit, 4132e78a4b19851d4483c1163fe1649da16728b4 it might be clearer)

tankyleo commented on lightning/src/ln/blinded_payment_tests.rs:
```diff
@@ -2368,10 +2609,43 @@ fn test_trampoline_unblinded_receive() {
+}
+
+#[test]
+fn test_trampoline_enforced_constrait_cltv() {
```
One more typo to squash here :)

tankyleo commented:
Also go ahead and squash next time you push, and expand the commit message to describe in detail what you are doing, including in the tests :) Would help me out with reviewing the code too thank you !

a-mpch commented on lightning/src/ln/blinded_payment_tests.rs:
```diff
@@ -2368,10 +2609,43 @@ fn test_trampoline_unblinded_receive() {
+		.without_claimable_event()
+		.expect_failure(HTLCHandlingFailureType::Receive { payment_hash });
+
 	do_pass_along_path(args);
```
My intention was to separate the big PR into smaller ones. This one seemed as a possibility to separate work but many of the code must be tested when the final feature of forwarding arrives.
That being said, in order to test the enforcement of the failed conditions, I modified the CLTV of the outer onion to something lower but acceptable (payment would be received) and maintaining the cltv_expiry of the receiving (that probably isn't clear, I'll modify it).
So in the case of no enforcement, payment is received. In case with enforcement it should fail because of the CLTV that the trampoline has is greater than the outer blinded path.

a-mpch commented on lightning/src/ln/blinded_payment_tests.rs:
```diff
@@ -2086,7 +2086,7 @@ fn do_test_trampoline_single_hop_receive(success: bool) {
 						pubkey: carol_node_id,
 						node_features: Features::empty(),
 						fee_msat: amt_msat,
-						cltv_expiry_delta: 24,
```
To calculate the outer onion of this receiving hop we do `cltv_expiry_delta` of the hop plus the current height 
In this case (24 + 32 = 56) 
Then when building the trampoline we do `excess_final_ctlv_expiry_delta` + current_height (39+32 = 72)

With the new constraints trampoline > outer onion. So it fails.

a-mpch commented:
@tankyleo Reading the initial PR needed a lot of explanation work 😅, thanks for the detailed review!

All your comments should have been addressed and rebased and squashed.

tankyleo commented on lightning/src/ln/blinded_payment_tests.rs:
```diff
@@ -2206,8 +2206,236 @@ fn test_trampoline_single_hop_receive() {
+		}
+		{
+		    let payment_failed_conditions = PaymentFailedConditions::new()
+				    .expected_htlc_error_data(LocalHTLCFailureReason::FinalIncorrectHTLCAmount, &[0, 0, 0, 0, 0, 0, 3, 232]);
```
@a-mpch Let's also do this here ? `let expected_error_data = amt_msat.to_be_bytes();`

tankyleo commented on lightning/src/ln/blinded_payment_tests.rs:
```diff
@@ -2206,8 +2206,237 @@ fn test_trampoline_single_hop_receive() {
+			do_commitment_signed_dance(&nodes[0], &nodes[1], &unblinded_node_updates.commitment_signed, false, false);
+		}
+		{
+		    let payment_failed_conditions = PaymentFailedConditions::new()
+				    .expected_htlc_error_data(LocalHTLCFailureReason::FinalIncorrectHTLCAmount, &[0, 0, 0, 0, 0, 0, 3, 232]);
+			  expect_payment_failed_conditions(&nodes[0], payment_hash, false, payment_failed_conditions);
```
nit: let's correct the indentation of line 2411 and 2413 :)

tankyleo commented on lightning/src/ln/blinded_payment_tests.rs:
```diff
@@ -2206,8 +2206,237 @@ fn test_trampoline_single_hop_receive() {
+}
+
+fn do_test_trampoline_unblinded_receive_constraint_failure(failure_scenario: TrampolineConstraintFailureScenarios) {
 	// Simulate a payment of A (0) -> B (1) -> C(Trampoline) (2)
```
Similar here, a quick little story of what the test does :) thank you very much !

tankyleo reviewed (commented):
Another round of review thank you !

Also proofread the commit message, there are a few mistakes in there :)

tankyleo commented on lightning/src/ln/onion_payment.rs:
```diff
@@ -306,6 +349,15 @@ pub(super) fn create_recv_pending_htlc_info(
 					}
 				})?;
 			let payment_data = msgs::FinalOnionHopData { payment_secret, total_msat };
+			check_trampoline_onion_constraints(outer_hop_data, cltv_expiry_height, sender_intended_htlc_amt_msat).map_err(|e| {
```
I understand that we don't have test coverage for the forwarding code paths, but here we don't have test coverage - should we be adding a test for blinded trampoline receives ?

a-mpch replied:
Yes, you are right. I'm adding it!

tankyleo commented on lightning/src/ln/onion_utils.rs:
```diff
@@ -1678,6 +1678,13 @@ pub enum LocalHTLCFailureReason {
 	HTLCMaximum,
 	/// The HTLC was failed because our remote peer is offline.
 	PeerOffline,
+	/// We have been unable to forward a payment to the next Trampoline node but may be able to
+	/// do it later.
+	TemporaryTrampolineFailure,
+	/// The amount or CLTV expiry were insufficient to route the payment to the next Trampoline.
+	TrampolineFeeOrExpiryInsufficient,
+	/// The specified next Trampoline node cannot be reached from our node.
+	UnknownNextTrampoline,
```
I would have expected these to be used in this PR since the commit message says 
```
We also add the expected errors by the spec: `TemporaryTrampolineFailure`,
`TrampolineFeeOrExpiryInsufficient` and `UnknownNextTrampoline` to be used
when encountering errors during forwarding or constraint validation.
```

But as far as I see they are not used anywhere yet, can you help me understand ? thank you again !

tankyleo commented on lightning/src/ln/blinded_payment_tests.rs:
```diff
@@ -2206,9 +2206,249 @@ fn test_trampoline_single_hop_receive() {
+	// Creates custom onion packet where the final trampoline hop uses unblinded receive format
+	// (not natively supported) to validate payment amount verification.
+	// - When underpay=false: Payment succeeds with correct amount
+	// - When underpay=true: Payment fails due to amount mismatch (sends 2x expected amount)
```
The blurb is great thank you ! If we send 2x the expected amount, shouldn't the boolean variable name be `overpay` ?

tankyleo commented on lightning/src/ln/blinded_payment_tests.rs:
```diff
@@ -2206,9 +2206,249 @@ fn test_trampoline_single_hop_receive() {
+					TrampolineHop {
+						pubkey: carol_node_id,
+						node_features: Features::empty(),
+						fee_msat: 0, // no trampoline fee becuase we are receiving.
```
nit: fix typo

tankyleo reviewed (commented):
Just two last questions, then a couple of drive-by nits again thank you for all the work.

tankyleo commented on lightning/src/ln/blinded_payment_tests.rs:
```diff
@@ -2206,9 +2206,249 @@ fn test_trampoline_single_hop_receive() {
+		}
+		{
+			let payment_failed_conditions = PaymentFailedConditions::new()
+				.expected_htlc_error_data(LocalHTLCFailureReason::FinalIncorrectHTLCAmount, &[0, 0, 0, 0, 0, 0, 3, 232]);
```
One more nit: let's use `let expected_error_data = amt_msat.to_be_bytes();` like we do below

a-mpch commented:
seems like a correct concern and approach. I'll take a look and base this PR on your branch!

a-mpch commented:
@valentinewallace I rebased in top of your branch.
I fixed tests on [35a0eb5](https://github.com/lightningdevkit/rust-lightning/pull/3983/commits/35a0eb58eba0c4bdae49918d7f10d8f09b868b74) adding an edge case for trampoline failures.
Let me know if it looks better!

a-mpch commented on lightning/src/ln/blinded_payment_tests.rs:
```diff
@@ -2535,6 +2535,6 @@ fn test_trampoline_forward_rejection() {
 		// Expect UnknownNextPeer error while we are unable to route forwarding Trampoline payments.
 		let payment_failed_conditions = PaymentFailedConditions::new()
 			.expected_htlc_error_data(LocalHTLCFailureReason::UnknownNextPeer, &[0; 0]);
-		expect_payment_failed_conditions(&nodes[0], payment_hash, false, payment_failed_conditions);
```
I had the same thought when doing this.
Alice built a route expecting the trampoline to find a path to the end route or to the next trampoline. But she selected a node that does not support trampoline forwarding.
Spec says: 
```
When processing a `trampoline_onion_packet`, a receiving node:

- If it doesn't support `trampoline_routing`:
  - MUST report a route failure to the origin node
```

We are actually raising `InvalidTrampolineForward` that has the same error-code that UnknownNextPeer. I didn't find another error code that makes sense. And `UnkownNextTrampoline` in spec is also a PERM error 🤔

a-mpch commented on lightning/src/ln/onion_utils.rs:
```diff
@@ -1390,6 +1373,16 @@ where
 			short_channel_id = route_hop.short_channel_id()
 		}
 
+		// If next hop is from a trampoline and there is only one trampoline hop
+		// in the trampoline route, means that the trampoline hop is also a
+		// final non-blinded node.
+		let is_next_hop_trampoline =
+			matches!(next_hop, Some((_, (Some(ErrorHop::TrampolineHop(..)), _))));
+		if is_next_hop_trampoline {
+			is_from_final_non_blinded_node =
+				path.blinded_tail.as_ref().map_or(0, |bt| bt.trampoline_hops.len()) == 1
+		}
```
I had the same idea but it would depend on the implementation of the recipient. [I left my question on the spec related to it](https://github.com/lightning/bolts/pull/836#discussion_r2349068696) as I couldn't find this case on eclair.

I think that adding the double wrap would make sense, but I have doubts as the trampoline wouldn't be an "intermediate", so maybe just need to behave as a common blinded route?

I'll check out and implement as it is done in the phantom case anyway!

valentinewallace replied:
Right... The recipient is a trampoline node in this case, so my thinking was that the suggested behavior would align with this part of the spec proposal:
https://github.com/lightning/bolts/pull/829/files#diff-c9739cca996bfa9ab52a39280655a99571ad6e17d51ffe2b2436d7fe2f68be80R225-R236
but it should be specified more in the BOLTs as well.

a-mpch replied:
https://github.com/lightningdevkit/rust-lightning/pull/4095

valentinewallace commented:
Need rebase 🎉 Thanks for the patience on this @a-mpch!
